File: src-lang/js/net/conn_redis.clj
(ns js.net.conn-redis
  (:require [hara.lang :as l]
            [std.lib.foundation :as f]
            [xt.lang.common-protocol :refer [defimpl.xt]])
  (:refer-clojure :exclude [print send]))

(l/script :js
  {:require [[xt.lang.spec-base :as xt]
             [xt.lang.spec-promise :as promise]
             [xt.lang.common-data :as xtd]
             [xt.lang.common-protocol :as protocol]
             [xt.net.conn-redis :as conn-redis]]
   :import  [["redis" :as [* Redis]]]})

(f/template-entries [l/tmpl-entry {:type :fragment
                                   :base "Redis"
                                   :tag "js"}]
  [createClient
   RedisClient
   print
   Multi
   AbortError
   RedisError
   ParserError
   ReplyError
   AggregateError
   addCommand])

(defn.js client-connect
  [client opts]
  (var #{defaults} client)
  (var env (xtd/obj-assign (xtd/obj-clone defaults) opts))
  (var url (+ "redis://"
              (or (xt/x:get-key env "host") "127.0.0.1")
              ":"
              (or (xt/x:get-key env "port") "6379")))
  (var raw (-/createClient {:url url}))
  (return
   (. (. raw (connect))
      (then (fn []
              (xt/x:set-key client "raw" raw)
              (return client))))))

(defn.js client-disconnect
  [client]
  (var #{raw} client)
  (return (. raw (quit))))

(defn.js client-exec
  [client command args]
  (var #{raw} client)
  (var input (xt/x:arr-assign [command] args))
  (return (. raw (sendCommand input))))

(defimpl.xt ^{:lang :js}
  JsRedisClient
  [defaults raw]
  conn-redis/IRedisClient
  {conn-redis/connect    -/client-connect
   conn-redis/disconnect -/client-disconnect
   conn-redis/exec       -/client-exec})

(defn.js create
  [defaults]
  (return
   (-/JsRedisClient (or defaults {}) nil)))

File: packages-gen/js/libs/xt-net/net/conn-redis.js
const Redis = require("redis")

const xtd = require("@xtalk/lang/common-data.js")

const conn_redis = require("@xtalk/net/conn-redis.js")

const protocol = require("@xtalk/lang/common-protocol.js")

function JsRedisClient(defaults,raw){
  globalThis["xt_lang_common_protocol$$IMPLEMENTATIONS"]["js.net.conn_redis/JsRedisClient"] = true;
  protocol.register_protocol_impl(conn_redis.IRedisClient["on"],"js.net.conn_redis/JsRedisClient",{
    "connect":client_connect,
    "disconnect":client_disconnect,
    "exec":client_exec
  });
  return {
    "::":"js.net.conn_redis/JsRedisClient",
    "::/protocols":[conn_redis.IRedisClient["on"]],
    "::/protocol-impls":{
        [conn_redis.IRedisClient["on"]]:{
          "connect":client_connect,
          "disconnect":client_disconnect,
          "exec":client_exec
        }
      },
    "defaults":defaults,
    "raw":raw
  };
}

function client_connect(client,opts){
  let {defaults} = client;
  let env = xtd.obj_assign(xtd.obj_clone(defaults),opts);
  let url = "redis://" + (env["host"] || "127.0.0.1") + ":" + (env["port"] || "6379");
  let raw = Redis.createClient({"url":url});
  return raw.connect().then(function (){
    client["raw"] = raw;
    return client;
  });
}

function client_disconnect(client){
  let {raw} = client;
  return raw.quit();
}

function client_exec(client,command,args){
  let {raw} = client;
  let input = xtd.arr_assign([command],args);
  return raw.sendCommand(input);
}

function create(defaults){
  return JsRedisClient(defaults || {},null);
}

module.exports = {
  ["client_connect"]:client_connect,
  ["client_disconnect"]:client_disconnect,
  ["client_exec"]:client_exec,
  ["JsRedisClient"]:JsRedisClient,
  ["create"]:create
}